### PR TITLE
Feature/sign in refactor

### LINKED ISSUE #4912: Implement magic links for sign in
Currently in progress on `feature/magic-sign-in`:

- [x] Call API to send magic link email (when implemented)
- [x] Call API to perform magic link auth (when implemented)
- [x] Handle 2FA for non-magic link login (WP.com)
- [x] Handle 2FA for self-hosted login
- [x] Show errors
- [x] Add field validation
- [x] Add analytics
- [x] UI Polish
- [x] Helpshift integration
- [x] 1Password integration
- [x] Fix animations on iPad
- [x] Allow back navigation through steps
- [x] Fix text formatting on password cell when showing password
- [ ] Update sms / authenticator message.
- [x] Handle special usernames like "admin"

### PR DESCRIPTION
**Pull request by aerych**
Closes #4912

This PR refactors the existing `LoginViewController` and implements a new "magic-links" sign-in feature.  The new controllers are written in Swift, and their UI is mainly defined in storyboard. View controllers are responsible for a single form, reducing the amount of state that needs to be maintained and tested.  Common functionality has been factored out into helper classes and protocols as much as possible, leaving controllers with a common, informal interface but no (or very limited) duplication. 

To test:
The `LoginViewController` supported a variety of sign-in scenarios, so there is a lot to test.  For each of the following scenarios, test the primary use-case, optional features on the form (various buttons) and error states (offline, invalid form entry). 

Scenario 1: 
Sign in to wpcom with an email address using the magic link flow.  To test the final step, reopening the app passing the auth token, you'll need to copy the `wordpress://` link from safari and change it to `wpdebug://`. 

Scenario 2: 
Sign in to wpcom with a username and password. No 2fa. 
 
Scenario 3: 
Sign in to wpcom with a username and password with 2fa.

Scenario 4: 
Sign in via a self-hosted blog. 

Scenario 5:
Confirm 1Password integration by testing on a device with 1Password installed.  Confirm that the 1Password button is present on the email, username/password, and self-hosted forms and that it functions on each. 
Note that on the self-hosted form you should be prompted to enter an URL if that field is empty. 

Scenario 6:
Confirm Safari stored password integration. 
Store credentials to wpcom 
- Settings > Safari > AutoFill > Names and Passwords.  
- Settings > Safari > Passwords > Add Password for wordpress.com
Launch and confirm you're prompted, and asked to use stored credentials. 
Choose credentials and confirm that they are filled out and you end up on the username/password form. 
Try again, and dismiss the prompt by clicking "Not Now".  Then re-open the prompt by clicking the saved passwords button on the Email form (first screen).

Needs review: @koke

### CODE REVIEW

File: WordPress/Classes/Extensions/UITextField+TextHelper.swift [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import UIKit
+
+extension UITextField
+{
+    func nonNilTrimmedText() -> String {
+        guard let str = text?.trim() else {
+            return ""
+        }
+        return str
+    }
```
Comment (koke) [inline WordPress/Classes/Extensions/UITextField+TextHelper.swift]: Interesting, I recently added `String.nonEmptyString` which is kind of the opposite approach: turn any empty string into a `nil` one. I find it a bit better to use optionals instead of empty strings as the compiler forces you to deal with the `nil` case in every occasion, but it's easy to forget handling an empty string.

File: WordPress/Classes/Networking/AccountServiceRemoteREST.m [hunk 2]
```diff
@@ -113,6 +117,72 @@ - (void)updateBlogsVisibility:(NSDictionary *)blogs
...
+        _anonymousApi = [WordPressComApi anonymousApi];
+    }
+
+    return _anonymousApi;
+}
+
+- (void)findExistingAccountByEmail:(NSString *)email success:(void (^)(BOOL found))success failure:(void (^)(NSError *error))failure
+{
+    // TODO: We need to make a versioned flavor of this endpoint and ensure it always returns a JSON object.
+    // TODO: Remove the special case in `WordPressComApi.assertApiVersion` the endpoint is versioned.
```
Comment (koke) [inline WordPress/Classes/Networking/AccountServiceRemoteREST.m]: We usually sign our `TODO`s with a name/date. Bonus points for a link to Github issue :wink:

File: WordPress/Classes/Networking/AccountServiceRemoteREST.m [hunk 3]
```diff
@@ -113,6 +117,72 @@ - (void)updateBlogsVisibility:(NSDictionary *)blogs
...
+                  NSDictionary *dict = (NSDictionary *)responseObject;
+                  NSString *errStr = [dict stringForKey:@"error"];
+                  found = [@"taken" isEqualToString:errStr];
+              }
+
+              if (success) {
+                  success(found);
+              }
+          } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
+              // The is-available endpoint is an oddball.
```
Comment (koke) [inline WordPress/Classes/Networking/AccountServiceRemoteREST.m]: Definitely. Until there's a newer version of this API, I'd add a comment explaining what the possible responses are. This code made no sense until I did a few test requests to that endpoint and saw the responses myself.

**Comment by kurzee** [comment]
Real quick, currently seeing a crash:
1. Sign in with a WordPress.com account.
2. On the My Sites tab > Switch Site > Hit the + to add a new site.
3. Select Add self-hosted site
4. Crash

```
'NSInvalidArgumentException', reason: 'NSConcreteAttributedString initWithString:: nil value'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010f0c7d85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x000000010eb3bdeb objc_exception_throw + 48
	2   CoreFoundation                      0x000000010f0c7cbd +[NSException raise:format:] + 205
	3   Foundation                          0x0000000109b1d18b -[NSConcreteAttributedString initWithString:] + 140
	4   Foundation                          0x0000000109b1d051 -[NSConcreteAttributedString initWithString:attributes:] + 27
	5   WordPress                           0x00000001077e1c4a -[WPWalkthroughTextField commonInit] + 1370
	6   WordPress                           0x00000001077e1432 -[WPWalkthroughTextField init] + 114
	7   WordPress                           0x00000001077e14a2 -[WPWalkthroughTextField initWithLeftViewImage:] + 66
	8   WordPress                           0x00000001077d4cbf -[LoginViewController addControls] + 2287
	9   WordPress                           0x00000001077d11c4 -[LoginViewController viewDidLoad] + 420
	10  UIKit                               0x000000010afad984 -[UIViewController loadViewIfRequired] + 1198
```

**Comment by aerych** [comment]
Nice catch @kurzee :)

File: WordPress/Classes/Networking/ServiceRemoteREST.h [hunk 4]
```diff
@@ -4,6 +4,7 @@ typedef enum
 {
     ServiceRemoteRESTApiVersion_1_1 = 1001,
     ServiceRemoteRESTApiVersion_1_2 = 1002,
+    ServiceRemoteRESTApiVersion_1_3 = 1003,
```
Comment (koke) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: Either I'm missing something or this doesn't seem to be used.

File: WordPress/WordPressApi/WordPressComApi.m [hunk 5]
```diff
@@ -158,7 +158,8 @@ - (void)assertApiVersion:(NSString *)URLString
 {
     NSAssert([URLString rangeOfString:@"/v1/"].length > 0
              || [URLString rangeOfString:@"v1.1"].length > 0
-             || [URLString rangeOfString:@"v1.2"].length > 0,
+             || [URLString rangeOfString:@"v1.2"].length > 0
+             || [URLString rangeOfString:@"is-available/email"].length > 0,
```
Comment (koke) [inline WordPress/WordPressApi/WordPressComApi.m]: If the v1.3 is actually used, it should be added here as well.

File: WordPress/Classes/ViewRelated/NUX/LoginFields.h [hunk 6]
```diff
@@ -11,6 +11,8 @@
 @property (nonatomic, copy) NSString *multifactorCode;
 @property (nonatomic, assign) BOOL userIsDotCom;
 @property (nonatomic, assign) BOOL shouldDisplayMultifactor;
+@property (nonatomic, assign) NSInteger safariStoredUsernameHash;
+@property (nonatomic, assign) NSInteger safariStoredPasswordHash;
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/LoginFields.h]: Probably a good chance to add nullability annotations to `LoginFields`, or even convert to Swift as it's basically a struct (even though it'd need to be a `class` for Obj-C compatibility)

**Comment by koke** [comment]
There's one thing that bugs me about the new design and I think it's making the self hosted flow more confusing.

When you land in the new flow, there's a single field asking for your username or email, and a next button. This suggests a sort of wizard and grabs your attention so it's easy to miss there's an "Add self hosted" link at the bottom. Once you tap on the text field, it will be the last time you see that option:
self-hosted-flow-fail

Compare this to the previous flow. You may not notice the link and enter your username and password, then get an error. Once you dismiss the error, you're back to the initial screen that you might scan more closely this time, noticing the self hosted option.

self-hosted-flow-previous

File: WordPress/Classes/ViewRelated/NUX/NUXAbstractViewController.swift [hunk 7]
```diff
@@ -0,0 +1,249 @@
+import UIKit
+import WordPressComAnalytics
+import WordPressShared
+
+/// A base class for the various NUX related related view controllers. 
+/// The base class sets up and configures common functionality, such as the help 
+/// button and badge.
+/// It is assumed that NUX controllers will always be presented modally.
+///
+class NUXAbstractViewController : UIViewController
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/NUXAbstractViewController.swift]: This looks like a common _chrome_ for all the NUX controllers. Did you consider making this a container instead of a base class?

File: WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import UIKit
+import WordPressShared
+
+let NUXSubmitButtonDisabledAlpha = CGFloat(0.25)
+
+/// A stylized button used by NUX controllers. The button presents white text 
+/// surrounded by a white border.  It also can display a `UIActivityIndicatorView`.
+///
+@objc class NUXSubmitButton : UIButton
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift]: This looks similar in functionality to `PurchaseButton`/`RoundedButton` although using a regular `UIActivityIndicatorView` instead of `MRActivityIndicatorView`. Maybe we could combine them and have the subclasses deal with just the styling.

**Comment by koke** [comment]
Some issues so far:

- The 1Password button from the email screen shows a list of unrelated logins
- From the self hosted screen with the keyboard up, if you drag to pop to the email controller there are some glitches on the 4s simulator. I don't see them on an iPad Air.
- In the Self Hosted login, the Next button on the (iPad) keyboard doesn't move to the next field for username/password
- When using an external keyboard, the forms still move as if there was a software keyboard on screen.
screen shot 2016-04-12 at 18 51 40
screen shot 2016-04-12 at 18 51 50

File: WordPress/Classes/Extensions/UITextField+TextHelper.swift (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Extensions/UITextField+TextHelper.swift]: I can see that, tho I think I prefer to avoid optionals where ever possible.  Its a nice helper method to reduce the amount of optional unwrapping that would otherwise be required.

File: WordPress/Classes/Networking/ServiceRemoteREST.h (hunk 4 above)
Comment (aerych) [inline WordPress/Classes/Networking/ServiceRemoteREST.h]: Oops. We ended up using v1 of the REST API for the endpoint, not v3.  Old code. Will purge.

File: WordPress/Classes/ViewRelated/NUX/NUXAbstractViewController.swift (hunk 7 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/NUXAbstractViewController.swift]: The original approach used parent/child view controllers.  It was a bit problematic in that we were maintaining more state than desired, wrangle a back button, and needing to manually manage the navigation history. We switching to a UINavigationController to simplify things and let it do the job it was intended for.

File: WordPress/Classes/ViewRelated/NUX/SigninWPComViewController.swift [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    @IBOutlet var bottomContentConstraint: NSLayoutConstraint!
+    @IBOutlet var verticalCenterConstraint: NSLayoutConstraint!
+    var onePasswordButton: UIButton!
+
+    var immediateSignin = false;
+
+    lazy var loginFacade: LoginFacade = {
+        let facade = LoginFacade()
+        facade.delegate = self
+        return facade
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninWPComViewController.swift]: I don't think you're setting the delegate to `nil` on `deinit` so this happens:

1. Launch the app and enter your username
2. Enter whatever you want in the password field
3. Tap the Sign In button and as soon as possible tap Back
4. You get back to the Email screen and get an error. When you dismiss the error the screen is white.

screen shot 2016-04-12 at 21 47 37

File: WordPress/Classes/ViewRelated/NUX/SigninWPComSyncHandler.swift [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+import WordPressComAnalytics
+
+/// A protocol and extension encapsulating syncing behavior common to WPCom 
+/// signin controllers.  Responsible for syncing account information and blog 
+/// details of the user.
+///
+protocol SigninWPComSyncHandler: class
+{
+    func configureLoading(loading: Bool)
+    func configureStatusMessage(message: String)
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninWPComSyncHandler.swift]: "configure" sounds odd when they just look like setters. I'd rather do `setStatusMessage(message: String)`. I'd prefer a property but not sure if that plays nice with ObjC `var statusMessage: String { get set }`

File: WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    /// proceeds with the submit action.
+    ///
+    func validateForm() {
+        view.endEditing(true)
+
+        // is reachable?
+        if !ReachabilityUtils.isInternetReachable() {
+            ReachabilityUtils.showAlertNoInternetConnection()
+            return
+        }
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: I recently learned from [a WWDC talk](https://developer.apple.com/videos/play/wwdc2012/706/) that Apple doesn't recommend using Reachability to *prevent* any connection. You should attempt the connection an if it fails, then use Reachability to determine when to retry.

File: WordPress/Classes/ViewRelated/NUX/SigninWPComViewController.swift (hunk 9 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninWPComViewController.swift]: Nice catch. I was able to reproduce this, thanks for the steps!  
I don't think it was the delegate.  The delegate has a weak reference so it shouldn't need to be set to nil. When testing it looked like the error vc was presented from the view controller that was being dismissed, which would leave things in a weird state for sure, but deinit wouldn't have been called at this point.  I've changed things so the error vc is instead presented from the nav controller. It would be better if LoginFacade had a halt/cancel/stop method we could call but that's out of scope here.

File: WordPress/Classes/ViewRelated/NUX/SigninWPComSyncHandler.swift (hunk 10 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninWPComSyncHandler.swift]: I'd prefer to keep with the "setupMethodName" and "configureMethodName" pattern used within the signin vcs. Rather than making them look more like setters how about we make them look less like setters.

File: WordPress/Classes/Networking/AccountServiceRemoteREST.m (hunk 3 above)
Comment (aerych) [inline WordPress/Classes/Networking/AccountServiceRemoteREST.m]: :+1: Comments added

File: WordPress/Classes/ViewRelated/NUX/LoginFields.h (hunk 6 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/LoginFields.h]: Macros added

File: WordPress/WordPressApi/WordPressComApi.m (hunk 5 above)
Comment (aerych) [inline WordPress/WordPressApi/WordPressComApi.m]: Nuked with fervor

**Comment by aerych** [comment]
>  - The 1Password button from the email screen shows a list of unrelated logins

Fixed.

>  - From the self hosted screen with the keyboard up, if you drag to pop to the email controller there are some glitches on the 4s simulator. I don't see them on an iPad Air.

I tested on a 5 and didn't see any glitches (sorry, no 4s here). I'm inclined to call it a simulator glitch but let's test a bit more first.

>  - In the Self Hosted login, the Next button on the (iPad) keyboard doesn't move to the next field for username/password

Fixed

>  - When using an external keyboard, the forms still move as if there was a software keyboard on screen. 

Fixed.

File: WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift (hunk 11 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: :+1:

Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: Reachability checks nuked.

File: WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift (hunk 8 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift]: I looked at this some.  They are functionally similar, but there doesn't look like a huge amount of overlap.  I'm not seeing where we'd gain much by extracting a common super class.

**Comment by aerych** [comment]
> I think it's making the self hosted flow more confusing.

I checked with @mattmiklic and we've added a button to the email/password from that will take the user to the self-hosted form.

**Comment by aerych** [comment]
Thanks Koke! Ready for another round.

File: WordPress/Classes/Extensions/UITextField+TextHelper.swift (hunk 1 above)
Comment (koke) [inline WordPress/Classes/Extensions/UITextField+TextHelper.swift]: On a second thought I think I agree with you. I'm not sure why `UITextField.text` is an optional, and adding to the problem, the header says the default is `nil` and the documentation says the default is `""`. It's probably good to coalesce the empty cases to "".

I'm not sure if I'd add trimming to the mix, but the function name is clear enough about it :+1:

File: WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift (hunk 8 above)
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/NUXSubmitButton.swift]: That's odd, I would have expected them to overlap more. We can leave it for now as the purchase is still a work in progress.

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+
+    /// Checks if the passed string matches a reserved username.
+    /// 
+    /// - Parameters:
+    ///     - username: The username to test.
+    ///
+    class func isUsernameReserved(username: String) -> Bool {
+        let name = username.lowercaseString.trim()
+        return ["admin", "administrator", "root"].contains(name)
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: This won't make a difference specially with such a short list, so feel free to ignore, but a `Set` is more adequate than an `Array` here. From the `NSSet` documentation:

> You can use sets as an alternative to arrays when the order of elements isn’t important and performance in testing whether an object is contained in the set is a consideration—while arrays are ordered, testing for membership is slower than with sets.

The possible performance gains here would be irrelevant, so this is more about using the right type than any practical improvement.

File: WordPress/Classes/Networking/AccountServiceRemoteREST.m (hunk 3 above)
Comment (koke) [inline WordPress/Classes/Networking/AccountServiceRemoteREST.m]: I've just realized that the API isn't as odd as I thought (aside from the non-JSON response), but we're negating the semantics here. What if the remote method matched the API more closely:

```objc
- (void)isEmailAvailable:(NSString *)email success:(void (^)(BOOL found))success failure:(void (^)(NSError *error))failure;
```

Then the service's `findExistingAccountByEmail` would call this and negate the response.

Do you think that'd make this easier to follow?

File: WordPress/Classes/System/WordPressAppDelegate.m [hunk 13]
```diff
@@ -434,6 +436,35 @@ - (void)application:(UIApplication *)application handleActionWithIdentifier:(NSS
...
+    [WPAppAnalytics track:WPAnalyticsStatLoginMagicLinkOpened];
+
+    // Show the signin view controller configured to perform the auth request.
+    [self showSigninScreen:token animated:NO thenEditor:NO];
+
+    return YES;
+}
+
+
+/**
```
Comment (koke) [inline WordPress/Classes/System/WordPressAppDelegate.m]: Can we move this inside `SignInHelpers` (or other helper class)?

**Comment by koke** [comment]
> To test the final step, reopening the app passing the auth token, you'll need to copy the wordpress:// link from safari and change it to wpdebug://.

I'm not being able to do that, Safari just fails and the address bar is empty. Any tips?

screen shot 2016-04-14 at 10 27 32

File: WordPress/Classes/System/WordPressAppDelegate.m [hunk 14]
```diff
@@ -518,8 +549,63 @@ - (void)showWelcomeScreenIfNeededAnimated:(BOOL)animated
...
+- (void)showSigninScreen:(NSString *)token animated:(BOOL)animated thenEditor:(BOOL)thenEditor
+{
+    // Complete a magic link signin if needed.
+    NSString *savedEmail = [SigninHelpers getEmailAddressForTokenAuth];
+    if (token && savedEmail) {
+        SigninLinkAuthViewController *controller = [SigninLinkAuthViewController controller:savedEmail token:token];
+        UINavigationController *navigationController = [[NUXNavigationController alloc] initWithRootViewController:controller];
+        UIViewController *rootViewController = self.window.rootViewController;
+        if (rootViewController.presentedViewController)  {
+            [rootViewController.presentedViewController dismissViewControllerAnimated:NO completion:^{
```
Comment (koke) [inline WordPress/Classes/System/WordPressAppDelegate.m]: Why do we need dismissing here? I fear there's a chance this could dismiss an unsaved post or comment reply. I haven't been able to test the magic link flow, and the new sign in only seems to be enabled for now for the first time setup. But I would expect this could make its way into all the others sign in flows, so my worry is that the following could happen:

1. You have a working self hosted blog in the app already
2. Try to log in to wp.com using the email link.
3. Email is a bit slow today so you give up for now and go back to replying to comments on your self hosted.
4. A notification pops up with new mail
5. Go to mail, tap the link, this is called, and your half written comment is lost

Another potential problematic flow:

1. You install the app and try to log in to wp.com using the magic link
2. Email is not arriving, so you decide to use the password instead
3. You start writing a post
4. The email finally arrives, and for some reason you tap the link
5. This is called, the editor is dismiss (I'd hope the content isn't lost though)
6. Not sure what else happens, but the rest looks safe

File: WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    func configureStatusLabel(message: String) {
+        statusLabel.text = message
+
+        sendCodeButton.hidden = !message.isEmpty
+    }
+
+
+    /// Configures the appearance and state of the submit button.
+    ///
+    func configureSubmitButton(animating: Bool) {
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: I'd make `animating` a named parameter (also in the other VCs that have this). When you see something like `configureSubmitButton(false)` it's really hard to know what the parameter means. `configureSubmitButton(animating: false)` is much more obvious.

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+            path = "http://\(path)"
+        }
+
+        let wpLogin = try! NSRegularExpression(pattern: "/wp-login.php$", options: .CaseInsensitive)
+        let wpadmin = try! NSRegularExpression(pattern: "/wp-admin/?$", options: .CaseInsensitive)
+        let trailingSlash = try! NSRegularExpression(pattern: "/?$", options: .CaseInsensitive)
+
+        path = wpLogin.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
+        path = wpadmin.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
+        path = trailingSlash.stringByReplacingMatchesInString(path, options: .ReportCompletion, range: NSRange(location: 0, length: path.characters.count), withTemplate: "")
```
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: Small (optional) suggestion, make a `String.dropSuffix(_)` helper

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift (hunk 12 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: A fair point but since I'd likely still initialize the set with an array literal I'll leave it as is for more compact code.

**Comment by aerych** [comment]
> Safari just fails and the address bar is empty. Any tips?

Try opening the email and tapping the link in the desktop Safari browser.  You should be able to copy and edit the `wordpress://` from the error prompt that appears.

File: WordPress/Classes/System/WordPressAppDelegate.m (hunk 14 above)
Comment (aerych) [inline WordPress/Classes/System/WordPressAppDelegate.m]: > Why do we need dismissing here?

If a UIViewController already has a presentedViewController it will not present another before the first is dismissed.   Its possible when the auth link is opened for the app is resumed vs launched in which case one of the other signin controllers would still be presented. 

Good thinking about the other uses cases. I can tweak the logic here.

File: WordPress/Classes/System/WordPressAppDelegate.m (hunk 13 above)
Comment (aerych) [inline WordPress/Classes/System/WordPressAppDelegate.m]: I wasn't sure which method this note was referring to so I made a best guess.  Let me know if I chose the wrong one.

File: WordPress/Classes/Networking/AccountServiceRemoteREST.m (hunk 3 above)
Comment (aerych) [inline WordPress/Classes/Networking/AccountServiceRemoteREST.m]: Maybe.  Let's try it.

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift (hunk 16 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: Adding helper if for no other reason than to have the utility going forward, but if you want to do a PR for that super cool NSURL extension you shared I'd be game to use that instead :)

**Comment by aerych** [comment]
I *think* that's handled everything for this round.  Tag @koke :)

File: WordPress/Classes/Extensions/String+Helpers.swift [hunk 17]
```diff
@@ -10,8 +10,24 @@ extension String {
...
     func nonEmptyString() -> String? {
         return isEmpty ? nil : self
     }
+
+    /// Remove the specified suffix from the string and returns the new string. 
+    ///
+    /// - Parameters: 
+    ///     - suffix: The suffxing to remove.  This can be a regular expression but should not include an ending `$`.
+    ///
+    func trimSuffix(suffix: String) -> String {
```
Comment (koke) [inline WordPress/Classes/Extensions/String+Helpers.swift]: The documentation does say that it expects a regex, but it can be easy to miss. I'd use a named parameter so it's obvious `string.trimSuffix(regexp: "wp-admin/?")`. Otherwise I'd expect it to do an exact string match

File: WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift (hunk 16 above)
Comment (koke) [inline WordPress/Classes/ViewRelated/NUX/SigninHelpers.swift]: :+1: We can leave that for another PR, since you mentioned this was just porting the existing Obj-C to Swift.

File: WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+    ///
+    func configureSendCodeButtonText() {
+        // Text: Verification Code SMS
+        let string = NSLocalizedString("Enter the code on your authenticator app or <u>send the code via text message</u>.",
+                                       comment: "Message displayed when a verification code is needed")
+        let options = [
+            NSDocumentTypeDocumentAttribute: NSHTMLTextDocumentType
+        ]
+
+        let styledString = "<style>body {font-family: sans-serif; font-size:14px; color: #ffffff; text-align:center;}</style>" + string
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: I've just been doing this same thing for a footer in Plans that has some highlighting. It looks like a font of `sans-serif` uses Helvetica instead of San Francisco. You'll need to use a value of `-apple-system` here if you want SF.

Comment (aerych) [inline WordPress/Classes/ViewRelated/NUX/Signin2FAViewController.swift]: Ah!  Thanks for the tip :)

**Comment by koke** [comment]
This is a small detail, and not sure if new, but:

1. Launch app
2. Enter wrong username, tap next
3. Enter password, tap sign in
4. Get an error
5. Tap on 1Password
6. Realize you were using the wrong username 
7. Tap on the right credentials
8. The app starts trying to sign in, but the username in the text field is still what you typed

If the 1Password credentials are good, it'll sign in correctly, but I found it a bit confusing, specially if the connection is a bit slow.

**Comment by koke** [comment]
Another fun bug:

1. Launch the app
2. Enter username, tap next
3. Enter password, tap sign in and immediately tap back
4. Enter another account username, tap next
5. Enter another account password, tap sign in

You're logged in as the first account. You can find the sites for both accounts in My Sites.

**Comment by aerych** [comment]
> The app starts trying to sign in, but the username in the text field is still what you typed

Good catch. 

> You're logged in as the first account. You can find the sites for both accounts in My Sites. 

The horror... the horror...

**Comment by kurzee** [comment]
>You're logged in as the first account. You can find the sites for both accounts in My Sites.

That's similar to this issue I've been looking at https://github.com/wordpress-mobile/WordPress-iOS/issues/5042. One problem here is that `AccountService` doesn't clean up the previous default account, so edge cases like this one end up creating two default accounts.

**Comment by aerych** [comment]
@koke, the bugs should be addressed.  Final check?

**Comment by koke** [comment]
:shipit: ❗ Really nice work here.

**Comment by aerych** [comment]
Thanks @koke !! :)
